DDF-1213 Improved query and caching performance under high load

pklinef opened the pull request:
- Made caching of non-cache mode queries optional and disabled it by
  default
- Switched to a fixed size thread pool
- Added batching to non-cache mode queries
- Refactored CacheCommitPhaser to encapsulate its implementation
  details to match CacheBatchAdder

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf-catalog/86)
<!-- Reviewable:end -->

pklinef commented:
@lessarderic @brendan-hofmann @jlcsmith

pklinef commented:
@lessarderic I am working on adding some unit tests for the code that was added but I wanted to get the review started in the meantime.

brendan-hofmann commented:
Review status: 6 of 7 files reviewed, 2 unresolved discussions, some commit checks pending.
Reviewed files:
- core/catalog-core-standardframework/pom.xml @ r2
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java @ r2
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SolrCache.java @ r1
- core/catalog-core-standardframework/src/main/resources/OSGI-INF/metatype/sortedFederationStrategy.xml @ r1
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r2
- docs/src/main/resources/Extending.adoc @ r1

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 45 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPyVntYlG4pNsStxy2)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L45)):</sup>
Would multi-threaded batching offer any performance gain?

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 105 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPy_WI6Cdm4QXhohPI)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L105)):</sup>
It looks like if this is called when the max backlog size is exceeded, the result list will be dropped on the floor. Is that right? Do we want that?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 5 of 7 files reviewed, 2 unresolved discussions, some commit checks failed.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 45 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPyVntYlG4pNsStxy2)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L45)):</sup>
It could and I thought about it but I was seeing a 20% increase in CPU usage with just a single thread due to all of the indexing when under load.  Caching metacards from endpoints is currently disabled by default given the resources needed if the system is under load.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 105 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPy_WI6Cdm4QXhohPI)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L105)):</sup>
Without multiple threads writing to a Solr cluster or dumping results to disk, it will not be possible to keep up under a worst case scenario such as a one or more clients crawling all results.  Caching metacards requested by endpoints is nice to have but not necessary.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Review status: all files reviewed, 2 unresolved discussions, some commit checks failed.
Reviewed files:
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r2
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r3

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 45 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPyVntYlG4pNsStxy2)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L45)):</sup>
That's too bad, glad you tried it though.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 105 \[r2\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqPy_WI6Cdm4QXhohPI)** ([raw file](https://github.com/codice/ddf-catalog/blob/db4d8e19046c06ccbaa8816f9fcbb8b2420913e3/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L105)):</sup>
Ok, that makes sense.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
@pklinef who's the hero for this PR?

---

Review status: all files reviewed, 2 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
@brendan-hofmann If you are not too busy, you can be the assignee.

brendan-hofmann commented:
Ok, I can do that.

---

Review status: all files reviewed, 2 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Everything passes, lgtm.

---

Review status: all files reviewed, 2 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 6 of 7 files reviewed, 7 unresolved discussions, some commit checks broke.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 42 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVo_dAoq1j5WK6QLI3)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L42)):</sup>
Should this be hard-coded or coming from a config property? Any chance we may want to change this value based on the system we're running on?

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 51 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVonKGhBGymNCcbFNo)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L51)):</sup>
Config property?

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 72 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVlBHdXQ5W-AW7eBYL)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L72)):</sup>
Why are we using metacardsToCache.values().size() everywhere instead of just metacardsToCache.size()?

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 90 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVmbEKkZy78WgWSLDS)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L90)):</sup>
Not sure we could get any exceptions while running this but should we have a catch RuntimeException and log something just in case?

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 104 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVoDQXj1FpAH-5yj-Z)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L104)):</sup>
Should we use some sort of configuration property for the 8 instead of hard-coding it here? I could see this value being dependent on the system we're running on.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 6 of 7 files reviewed, 8 unresolved discussions, some commit checks broke.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 376 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVsHXVREroyesiwyoO)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L376)):</sup>
Should we catch and log RuntimeExceptions here? Do you know if the thread dies and is removed from the pool if an exception occurs?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 6 of 7 files reviewed, 8 unresolved discussions, some commit checks broke.

---

<sup>**[core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java, line 53 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqWA6y91TAubuMT1mpy)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java#L53)):</sup>
We should verify that the collection passed to create() is what we expect instead of just using anyCollectionOf(), otherwise we're missing an important part of the test.

---

<sup>**[core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java, line 55 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqWAqpY_R8y2wl82k1E)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java#L55)):</sup>
We should also test the following scenarios:

- Empty batch
- Batch that exceeds the max backlog size
- Invalid inputs (e.g., null or negative values)
- Exception thrown inside the Runnables
- Calling add with a few metacards repeated to test cache updates

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: all files reviewed, 8 unresolved discussions, some commit checks broke.
Reviewed files:
- core/catalog-core-standardframework/pom.xml @ r2
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r2
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java @ r2
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SolrCache.java @ r1
- core/catalog-core-standardframework/src/main/resources/OSGI-INF/metatype/sortedFederationStrategy.xml @ r1
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r4
- docs/src/main/resources/Extending.adoc @ r1

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 4 of 7 files reviewed, 8 unresolved discussions, all commit checks successful.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 42 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVo_dAoq1j5WK6QLI3)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L42)):</sup>
I thought about it for backlog size and batch size but I do not want to overwhelm the the federation strategy configuration with too many configuration options.  These values should be good enough for most users.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 51 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVonKGhBGymNCcbFNo)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L51)):</sup>
See reply on backlog size.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 72 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVlBHdXQ5W-AW7eBYL)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L72)):</sup>
I have no idea what steps I went through to write that and leave it but that was a mistake.  Fixed.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 90 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVmbEKkZy78WgWSLDS)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L90)):</sup>
Good catch.  Fixed.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 104 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVoDQXj1FpAH-5yj-Z)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L104)):</sup>
I thought about it but it would need to be a constructor argument to avoid having to stop adding to the current executor service before shutting it down and creating a new one.  I did not want to add configuration that would restart the federation strategy as requests are actively being processed.  We could use a system property or new etc file but that is not a pattern we have used and would need to consider the tradeoffs.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 376 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVsHXVREroyesiwyoO)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L376)):</sup>
If an exception happened, it would be in the Future that is returned from the submit call.  This caching is best effort so it should not be a problem if it fails.

---

<sup>**[core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java, line 53 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqWA6y91TAubuMT1mpy)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java#L53)):</sup>
Fixed.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 6 of 7 files reviewed, 2 unresolved discussions, some commit checks broke.
Reviewed files:
- core/catalog-core-standardframework/pom.xml @ r5
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r5

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 376 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVsHXVREroyesiwyoO)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L376)):</sup>
My worry was that the thread would die and get removed from the thread pool if an exception was thrown but I couldn't find any clear answer on that, so up to you.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 6 of 7 files reviewed, 2 unresolved discussions, some commit checks broke.

---

<sup>**[core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java, line 55 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqWAqpY_R8y2wl82k1E)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java#L55)):</sup>
What did you have in mind for empty batch?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 5 of 7 files reviewed, 2 unresolved discussions, some commit checks pending.

---

<sup>**[core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java, line 55 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqWAqpY_R8y2wl82k1E)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java#L55)):</sup>
Add an empty result, wait for the flush time and make sure the cache isn't called. Also, in the normal test case, should would wait for a second flush time and make sure the cache isn't called again just to verify that the batch is really getting cleared?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 5 of 7 files reviewed, 3 unresolved discussions, some commit checks pending.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 52 \[r6\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqX2RwhECy5G17ydS3_)** ([raw file](https://github.com/codice/ddf-catalog/blob/d82a21d13b475f95679bc292ca4a8144ae94582f/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L52)):</sup>
I liked maxBacklogSize better :)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: all files reviewed, 3 unresolved discussions, some commit checks pending.
Reviewed files:
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r6
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r6

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 5 of 7 files reviewed, 2 unresolved discussions, some commit checks pending.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java, line 52 \[r6\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqX2RwhECy5G17ydS3_)** ([raw file](https://github.com/codice/ddf-catalog/blob/d82a21d13b475f95679bc292ca4a8144ae94582f/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java#L52)):</sup>
Good point, I like it better too. Fixed. :)

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 376 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVsHXVREroyesiwyoO)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L376)):</sup>
Since I submitted it as a task, the AbstractExecutorService will wrap the Runnable as a FutureTask and the FutureTask will catch all Throwables and record them so it can throw an ExecutionException when get() is called on the returned future.  So the thread will not be removed from the thread pool.

https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/util/concurrent/AbstractExecutorService.java#L73-L134

https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/util/concurrent/FutureTask.java#L333-L338

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 4 of 7 files reviewed, 2 unresolved discussions, some commit checks pending.

---

<sup>**[core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java, line 376 \[r4\]](https://reviewable.io:443/reviews/codice/ddf-catalog/86#-JqVsHXVREroyesiwyoO)** ([raw file](https://github.com/codice/ddf-catalog/blob/f8a0c0a4e817adbe5e6927498b1d6e7e42baff3d/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java#L376)):</sup>
I went ahead and added logging just in case so it will be easier to catch a bug in case the unit tests are missing something or more code is added that might not be as well unit tested.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: all files reviewed, all discussions resolved, some commit checks failed.
Reviewed files:
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r7
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java @ r7
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r7

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Review status: all files reviewed, all discussions resolved, some commit checks failed.
Reviewed files:
- core/catalog-core-standardframework/pom.xml @ r5
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CacheBulkProcessor.java @ r7
- core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/CachingFederationStrategy.java @ r7
- core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CacheBulkProcessorTest.java @ r7

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-catalog/86)
<!-- Sent from Reviewable.io -->
